core/templates/span.h

```diff
@@ -0,0 +1,114 @@
+/**************************************************************************/
+/*  span.h                                                                */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef SPAN_H
+#define SPAN_H
+
+#include "core/typedefs.h"
+
+// Equivalent of std::span.
+// Represents a view into a contiguous memory space.
+// DISCLAIMER: This data type does not own the underlying buffer. DO NOT STORE IT.
+//  Additionally, for the lifetime of the Span, do not add elements to or remove from the buffer.
+//  Failure to respect this may lead to crashes or undefined behavior.
+template <typename T>
+class Span {
+	T *_ptr = nullptr;
+	size_t _len = 0;
+
+public:
+	static constexpr bool is_mutable = std::negation_v<std::is_const<T>>;
+
+	Span() = default;
+	Span(T *p_ptr, size_t p_len) :
+			_ptr(p_ptr), _len(p_len) {}
```

Comment:
If you wanna get *really* fancy, you can make the length argument implicit:
```suggestion
	Span() = default;
	Span(T *p_ptr, size_t p_len) :
			_ptr(p_ptr), _len(p_len) {}
	template <size_t len>
	Span(T (&p_ptr)[len]) :
			_ptr(p_ptr), _len(len) {}
```

Reply:
This would unfortunately be incorrect for string types, since the `NULL` terminator is part of the `[len]`. I may add a follow up PR to add array literal constructors, and specialize string literal constructors to use `strlen()`, but I thought it too controversial to end up in the initial PR.